adcc/LazyMp.py

```diff
@@ -60,39 +57,80 @@ def __getattr__(self, attr):
-        amplitudes in this class. Note, that potential other caches, such as
-        computed ADC intermediates are *not* automatically invalidated.
-        """
-        return self.__cpp.set_t2(space, tensor)
+        hf = self.reference_state
+        if space == b.oovv:
+            eia = self.df(b.ov)
+            return compute_t2(hf.oovv, eia, eia)
+        elif space == b.ocvv:
+            eia = self.df(b.ov)
+            eca = self.df(b.cv)
+            return compute_t2(hf.ocvv, eia, eca)
+        elif space == b.ccvv:
+            eca = self.df(b.cv)
+            return compute_t2(hf.ccvv, eca, eca)
+        else:
+            raise NotImplementedError("T2 amplitudes not "
+                                      f"implemented for space {space}.")
 
+    @cached_member_function
     def td2(self, space):
         """Return the T^D_2 term"""
-        return self.__cpp.td2(space)
+        if space != b.oovv:
+            raise NotImplementedError("T^D_2 term not implemented "
+                                      f"for space {space}.")
+        t2erit = self.t2eri(b.oovv, b.ov).transpose((1, 0, 2, 3))
+        denom = direct_sum(
+            'ia,jb->ijab', self.df(b.ov), self.df(b.ov)
+        ).symmetrise(0, 1)
+        return evaluate(
+            (+ 4.0 * t2erit.antisymmetrise(2, 3).antisymmetrise(0, 1)
+             - 0.5 * self.t2eri(b.oovv, b.vv)
+             - 0.5 * self.t2eri(b.oovv,  b.oo)) / denom
+        )
 
+    @cached_member_function
     def t2eri(self, space, contraction):
         """
         Return the T2 tensor with ERI tensor contraction intermediates.
         These are called pi1 to pi7 in libadc.
         """
-        return self.__cpp.t2eri(space, contraction)
+        hf = self.reference_state
+        expressions = {
+            # space + contraction
+            b.ooov + b.vv: einsum('ijbc,kabc->ijka', self.t2oo, hf.ovvv),
+            b.ooov + b.ov: einsum('ilab,lkjb->ijka', self.t2oo, hf.ooov),
+            b.oovv + b.oo: einsum('klab,ijkl->ijab', self.t2oo, hf.oooo),
+            b.oovv + b.ov: einsum('jkac,kbic->ijab', self.t2oo, hf.ovov),
+            b.oovv + b.vv: einsum('ijcd,abcd->ijab', self.t2oo, hf.vvvv),
+            b.ovvv + b.oo: einsum('jkbc,jkia->iabc', self.t2oo, hf.ooov),
+            b.ovvv + b.ov: einsum('ijbd,jcad->iabc', self.t2oo, hf.ovvv),
+        }
+        return expressions[space + contraction].evaluate()
 
-    @property
+    @cached_property
+    @timed_member_call(timer="timer")
     def mp2_diffdm(self):
         """
         Return the MP2 differensce density in the MO basis.
         """
-        ret = OneParticleOperator.from_cpp(self.__cpp.mp2_diffdm)
+        hf = self.reference_state
+        ret = OneParticleOperator(self.mospaces, is_symmetric=True)
+        blocks = ["oo", "ov", "vv"]
+        if self.has_core_occupied_space:
+            blocks += ["cc", "co", "cv"]
+        for bl in blocks:
+            ba = getattr(b, bl)
```

Comment:
I'm debating with myself whether being explicit and having the list in `o1o1` etc. is better or not.

Reply:
Also resolved.